src/economy/shops/shop.ts

```diff
@@ -0,0 +1,389 @@
+import { color, text, tooltip, translate } from 'craftjs-plugin/chat';
+import { UUID } from 'java.util';
+import { TranslatableComponent } from 'net.md_5.bungee.api.chat';
+import {
+  Bukkit,
+  ChatColor,
+  Location,
+  Material,
+  SoundCategory,
+} from 'org.bukkit';
+import { Block, Chest, Container } from 'org.bukkit.block';
+import { Player } from 'org.bukkit.entity';
+import { Action as BlockAction } from 'org.bukkit.event.block';
+import { PlayerInteractEvent } from 'org.bukkit.event.player';
+import { EquipmentSlot, ItemStack } from 'org.bukkit.inventory';
+import { ChatMessage, GLOBAL_PIPELINE } from '../../chat/pipeline';
+import { addItemTo, giveItem } from '../../common/helpers/inventory';
+import { getItemName } from '../../common/helpers/items';
+import { Currency, getShopCurrency } from '../currency';
+import { getInventoryBalance, giveMoney, takeMoneyFrom } from '../money';
+import { findItemsFromInventory, getBlockBehind, getShopItem } from './helpers';
+import { openShopGUI } from './shop-gui';
+import { getShop } from './ShopData';
+import { distanceBetween } from '../../common/helpers/locations';
+import { errorMessage } from '../../chat/system';
+import { getTaxes, sendTaxes } from './taxes';
+import { round } from '../../common/helpers/math';
+
+registerEvent(PlayerInteractEvent, (event) => {
+  if (event.action !== BlockAction.RIGHT_CLICK_BLOCK) return;
+  if (event.hand !== EquipmentSlot.HAND) return;
+  if (!event.clickedBlock) return;
+  const sign = event.clickedBlock;
+  const view = getShop(sign);
+
+  if (!view) return;
+  if (!view.type) return;
+
+  if (event.player.isSneaking()) {
+    if (event.isBlockInHand()) return;
+    openShopGUI(event.player, sign);
+  }
+  displayShopInfo(event.player, sign);
+});
+
+function displayShopInfo(p: Player, sign: Block) {
+  const view = getShop(sign);
+  if (!view) return;
+  const item = getShopItem(view.item);
+  const chestBlock = getBlockBehind(sign);
+  if (!(chestBlock?.state instanceof Chest)) return;
+  const chest = chestBlock.state;
+  if (!item) return;
+  const taxes = getTaxes(view.taxRate, view.price);
+  const taxFreePrice = view.price - taxes;
+  const unit =
+    view.price === 1
+      ? view.currency.unitPlural.slice(0, -1)
```

Comment:
I think this will not produce correct singular representation for many Finnish (and some English) words.